ledger/accountdb.go

```diff
@@ -633,6 +644,414 @@ func accountsAddNormalizedBalance(tx *sql.Tx, proto config.ConsensusParams) erro
+		VoteID:             ba.VoteID,
+		SelectionID:        ba.SelectionID,
+		VoteFirstValid:     ba.VoteFirstValid,
+		VoteLastValid:      ba.VoteLastValid,
+		VoteKeyDilution:    ba.VoteKeyDilution,
+		AuthAddr:           ba.AuthAddr,
+		TotalAppSchema: basics.StateSchema{
+			NumUint:      ba.TotalAppSchemaNumUint,
+			NumByteSlice: ba.TotalAppSchemaNumByteSlice,
+		},
+		TotalExtraAppPages: ba.TotalExtraAppPages,
+	}
+}
+
+type resourceFlags uint8
+
+const (
+	resourceFlagsHolding    = 0 //nolint:deadcode,varcheck
+	resourceFlagsNotHolding = 1
+	resourceFlagsOwnership  = 2
+)
+
+type resourcesData struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	// asset parameters ( basics.AssetParams )
+	Total         uint64         `codec:"a"`
+	Decimals      uint32         `codec:"b"`
+	DefaultFrozen bool           `codec:"c"`
+	UnitName      string         `codec:"d"`
+	AssetName     string         `codec:"e"`
+	URL           string         `codec:"f"`
+	MetadataHash  [32]byte       `codec:"g"`
+	Manager       basics.Address `codec:"h"`
+	Reserve       basics.Address `codec:"i"`
+	Freeze        basics.Address `codec:"j"`
+	Clawback      basics.Address `codec:"k"`
+
+	// asset holding ( basics.AssetHolding )
+	Amount uint64 `codec:"l"`
+	Frozen bool   `codec:"m"`
+
+	// application local state ( basics.AppLocalState )
+	SchemaNumUint      uint64              `codec:"n"`
+	SchemaNumByteSlice uint64              `codec:"o"`
+	KeyValue           basics.TealKeyValue `codec:"p"`
+
+	// application global params ( basics.AppParams )
+	ApprovalProgram               []byte              `codec:"q,allocbound=config.MaxAvailableAppProgramLen"`
+	ClearStateProgram             []byte              `codec:"r,allocbound=config.MaxAvailableAppProgramLen"`
+	GlobalState                   basics.TealKeyValue `codec:"s"`
+	LocalStateSchemaNumUint       uint64              `codec:"t"`
+	LocalStateSchemaNumByteSlice  uint64              `codec:"u"`
+	GlobalStateSchemaNumUint      uint64              `codec:"v"`
+	GlobalStateSchemaNumByteSlice uint64              `codec:"w"`
+	ExtraProgramPages             uint32              `codec:"x"`
+
+	// ResourceFlags helps to identify which portions of this structure should be used; in particular, it
+	// helps to provide a marker - i.e. whether the account was, for instance, opted-in for the asset compared
+	// to just being the owner of the asset. A comparison against the empty structure doesn't work here -
+	// since both the holdings and the parameters are allowed to be all at their default values.
+	ResourceFlags resourceFlags `codec:"y"`
+	UpdateRound   uint64        `codec:"z"`
+}
+
+func (rd *resourcesData) SetAssetParams(ap basics.AssetParams, haveHoldings bool) {
+	rd.Total = ap.Total
+	rd.Decimals = ap.Decimals
+	rd.DefaultFrozen = ap.DefaultFrozen
+	rd.UnitName = ap.UnitName
+	rd.AssetName = ap.AssetName
+	rd.URL = ap.URL
+	copy(rd.MetadataHash[:], ap.MetadataHash[:])
```

Comment:
why copy this one but not the byte arrays inside addresses like Manager, Reserve, etc, or the slices like ApprovalProgram?

Reply:
I wasn't sure if it would copy byte arrays - but apparently the answer is "yes", so I'll change that over.